### PR TITLE
test: add test for .ts files processing

### PR DESCRIPTION
**Pull request by antongolub**
* Replace deprecated **tsc** pkg with typescript's compiler invocation
* Add .ts files processing test.

- [x] Tests pass

### CODE REVIEW

**Comment by antongolub** [comment]
WIP, some issue found

File: zx.mjs [hunk 1]
```diff
@@ -117,7 +118,8 @@ async function importPath(filepath, origin = filepath) {
   if (ext === '.ts') {
     let {dir, name} = parse(filepath)
     let outFile = join(dir, name + '.mjs')
-    await run`npm_config_yes=true npx tsc --target esnext --module esnext --moduleResolution node ${filepath}`
+    let tsc = await which('tsc').then(() => 'tsc' ).catch(() => 'npm_config_yes=true npx tsc')
```
Comment (antonmedv) [inline zx.mjs:121]: I think nox will do it automatically?

Comment (antongolub) [reply zx.mjs:121]: `npx <...>` is not equal `npm i -g && node <...>`. You may check this with `--verbose` flag.

Comment (antongolub) [reply zx.mjs]: Binary name and pkg name may not match in general case. So it `npx` performs request every time.

File: zx.mjs [hunk 2]
```diff
@@ -117,7 +118,8 @@ async function importPath(filepath, origin = filepath) {
   if (ext === '.ts') {
     let {dir, name} = parse(filepath)
     let outFile = join(dir, name + '.mjs')
-    await run`npm_config_yes=true npx tsc --target esnext --module esnext --moduleResolution node ${filepath}`
+    let tsc = await which('tsc').then(() => 'tsc', () => 'npm_config_yes=true npx tsc')
```
Comment (antonmedv) [inline zx.mjs:121]: What if user want to run tsc from node_nodules and have globally installed tsc?

File: zx.mjs [hunk 3]
```diff
@@ -173,8 +175,11 @@ function transformMarkdown(source) {
 
 async function run(pieces, ...args) {
   let v = $.verbose
+  let q = $.quote
   $.verbose = false
+  $.quote = v => v
```
Comment (antonmedv) [inline zx.mjs:180]: Why this?

Comment (antongolub) [reply zx.mjs:180]: `$.quote` adds redundant quotes on template render: `"`npm_config_yes=true npx tsc` --module ..."`

File: zx.mjs (hunk 2 above)
Comment (antongolub) [reply zx.mjs:121]: We may introduce `$.tsc` for this case.

Comment (antongolub) [reply zx.mjs:121]: Not sure, but this case might be handled by `npx`. Afair, npx tries to find pkg locally, and if it's absent, then fetches from registry.

Comment (antongolub) [reply zx.mjs:121]: We can check `node ./node_modules/.bin/tsc`.

File: zx.mjs (hunk 3 above)
Comment (antongolub) [reply zx.mjs]: Reverted.

**Comment by antongolub** [comment]
Everything seems fine now, rfr.

File: zx.mjs (hunk 1 above)
Comment (antonmedv) [reply zx.mjs]: But for tsc it’s not a concern, right?

File: package.json [hunk 4]
```diff
@@ -11,7 +11,9 @@
     "node": ">= 14.8.0"
   },
   "scripts": {
-    "test": "node zx.mjs test.mjs"
+    "test": "npm run test:mjs && npm run test:ts",
+    "test:mjs": "node zx.mjs test.mjs",
```
Comment (antonmedv) [inline package.json:15]: Let’s keep only one test. Why to separate?

File: test-tsc.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
+// Copyright 2021 Google LLC
```
Comment (antonmedv) [inline test-tsc.ts:1]: Let’s just import test from examples dir. see other tests.

File: zx.mjs [hunk 6]
```diff
@@ -117,7 +117,7 @@ async function importPath(filepath, origin = filepath) {
   if (ext === '.ts') {
     let {dir, name} = parse(filepath)
     let outFile = join(dir, name + '.mjs')
-    await run`npm_config_yes=true npx tsc --target esnext --module esnext --moduleResolution node ${filepath}`
+    await run`npm_config_yes=true npx -p typescript tsc --target esnext --module esnext --moduleResolution node ${filepath}`
```
Comment (antonmedv) [inline zx.mjs:120]: Tsc comes with ts by default?

Comment (antongolub) [reply zx.mjs:120]: Yes. https://github.com/microsoft/TypeScript/blob/master/package.json#L24

Comment (antonmedv) [reply zx.mjs:120]: Why do we need to specify it here?

File: package.json (hunk 4 above)
Comment (antongolub) [reply package.json:15]: Just for atomicity.

File: test-tsc.ts (hunk 5 above)
Comment (antongolub) [reply test-tsc.ts:1]: I suggest reusing `test.mjs` instead.

File: zx.mjs (hunk 6 above)
Comment (antongolub) [reply zx.mjs:120]: To load **typescript** pkg and then invoke its **tsc** bin.

Comment (antonmedv) [reply zx.mjs:120]: I think npx will work without it too. Correct?

Comment (antongolub) [reply zx.mjs:120]: If pkg name does not equal to its bin we have to specify what exactly we'd like to invoke.

File: test-tsc.ts (hunk 5 above)
Comment (antongolub) [reply test-tsc.ts]: How do you like this trick?

File: zx.mjs (hunk 6 above)
Comment (antonmedv) [reply zx.mjs:120]: I got it. https://www.npmjs.com/package/tsc is deprecated.
I assumed we using an old tsc package.

Comment (antonmedv) [inline zx.mjs:120]: Let's separate this change from other test-related changes.

**Review by antonmedv — CHANGES_REQUESTED** [review]

File: zx.mjs (hunk 6 above)
Comment (antongolub) [reply zx.mjs:120]: https://github.com/google/zx/pull/117

**Comment by antongolub** [comment]
Rebased and squashed

File: index.d.ts [hunk 7]
```diff
@@ -48,16 +48,31 @@ export class ProcessOutput {
...
-  export const nothrow: (p: ProcessPromise<ProcessOutput>) => ProcessPromise<ProcessOutput>
-  export const sleep: (ms: number) => Promise<void>
-  export const question: (query?: string, options?: QuestionOptions) => Promise<string>
-  export const chalk: typeof _chalk
-  export const fs: typeof _fs & {
-    createWriteStream: typeof createWriteStream
-    createReadStream: typeof createReadStream
-  }
-  export const os: typeof _os
+  const $: $
```
Comment (antonmedv) [inline index.d.ts:70]: I checked and it seems what we can expert from declare global. This is better as no duplicates.

Comment (antongolub) [reply index.d.ts:70]: And the test is still passing?

Comment (antonmedv) [reply index.d.ts:70]: Yes, you’re right.

**Review by antonmedv — APPROVED** [review]

**Comment by antongolub** [comment]
@antonmedv, what's wrong with ts-test, why it was reverted?

**Comment by antonmedv** [comment]
I thought it’s not good to have one file with ts and mjs.

**Comment by antongolub** [comment]
This approach brought zero code duplication)
1. Why we should not check named exports?
2. Should we have TS tests, if its support is declared?

**Comment by antonmedv** [comment]
What I was thinking is to test not only declarations but loader too.
```js
{ // Markdown scripts are working
  await $`node zx.mjs examples/markdown.md`
}

{ // TypeScript scripts are working
  await $`node zx.mjs examples/typescript.ts`
}
```
☝️ I think this is enough for our case 😁
What do you think?

**Comment by antongolub** [comment]
My prev test had used loader as well))
Anyway, I agree, let's add some tests for **ts** and **md** files as you suggested. https://github.com/google/zx/pull/121

**Pull request merged**